### DISCUSSION TITLE
Does meson's own syntax support recursively searching for all .c files within a folder?

### DISCUSSION THREAD

**Discussion by AnlangA** [mesonbuild/meson#13353] [Q&A]
Note that this is without using Python or any other scripting language to assist meson.

**Comment by AnlangA** [comment]
Why can't I specify target files with a wildcard?
Instead of specifying files explicitly, people seem to want to do this:

executable('myprog', sources : '*.cpp') # This does NOT work!
Meson does not support this syntax and the reason for this is simple. This cannot be made both reliable and fast. By reliable we mean that if the user adds a new source file to the subdirectory, Meson should detect that and make it part of the build automatically.

One of the main requirements of Meson is that it must be fast. This means that a no-op build in a tree of 10 000 source files must take no more than a fraction of a second. This is only possible because Meson knows the exact list of files to check. If any target is specified as a wildcard glob, this is no longer possible. Meson would need to re-evaluate the glob every time and compare the list of files produced against the previous list. This means inspecting the entire source tree (because the glob pattern could be src/\*/\*/\*/\*.cpp or something like that). This is impossible to do efficiently.

The main backend of Meson is Ninja, which does not support wildcard matches either, and for the same reasons.

Because of this, all source files must be specified explicitly.

**Comment by Ekmansoft** [comment]
This is a decision that has been made, but not necessarily a good one.

**Reply by eli-schwartz** [reply]
You can put the list of sources into a text file such as `sources.lst` and read it into meson with `source_files = import('fs').read('sources.lst').split()`,  which makes it easy to maintain the list via e.g. a git hook that runs `find` and prints to the file containing the list.

Or as noted, if you're really positive you want globbing and are willing to accept the tradeoffs, you can use a python script to glob for you.
